### PR TITLE
Add a new module ce_lldp to manage Link Layer Discovery Protocol(LLDP)

### PR DESCRIPTION
**Pull request by yuandongx**
##### SUMMARY

A new module ce_lldp to manage Link Layer Discovery Protocol(LLDP)

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/network/cloudengine/ce_lldp.py
##### ADDITIONAL INFORMATION

```paste below
```
### task
```
  - name: "test the correct lldpenable"
    ce_lldp:
      lldpenable: enabled
      mdnstatus: rxOnly


```
### log info
```

changed: [10.190.121.125] => {
    "changed": true,
    "end_state": {
        " bind_name": " None",
        " fast_count": " 4",
        " hold_multiplier": " 4",
        " interval": " 30",
        " management_address": " None",
        " mdn_notification_interval": " 5",
        " mdnstatus": " rxOnly",
        " notification_interval": " 5",
        " restart_delay": " 2",
        " transmit_delay": " 2",
        "lldpenable": " enabled"
    },
    "existing": {
        " mdnstatus": " disabled",
        "lldpenable": " disabled"
    },
    "invocation": {
        "module_args": {
            "bind_name": null,
            "fast_count": null,
            "hold_multiplier": null,
            "interval": null,
            "lldpenable": "enabled",
            "management_address": null,
            "mdn_notification_interval": null,
            "mdnstatus": "rxOnly",
            "notification_interval": null,
            "restart_delay": null,
            "state": "present",
            "timeout": null,
            "transmit_delay": null,
        }
    },
    "proposed": {
        "lldpenable": "enabled",
        "mdnstatus": "rxOnly"
    },
    "updates": [
        "lldp enable lldp mdn",
         "enable lldp mdn enable"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
### show on device
```
[~8850_130.12]dis cu | in lldp
lldp enable
lldp mdn enable
[~8850_130.12]
```
### task
```

  - name: "test the correct interval"
    ce_lldp:
      interval: 8
      hold_multiplier: 2
      transmit_delay: 1
      restart_delay: 10
      notification_interval: 10
      fast_count: 3
      mdn_notification_interval: 10

```
### test log
```

changed: [10.190.121.125] => {
    "changed": true,
    "end_state": {
        " bind_name": " None",
        " fast_count": " 3",
        " hold_multiplier": " 2",
        " interval": " 8",
        " management_address": " None",
        " mdn_notification_interval": " 10",
        " mdnstatus": " rxOnly",
        " notification_interval": " 10",
        " restart_delay": " 10",
        " transmit_delay": " 1",
        "lldpenable": " enabled"
    },
    "existing": {
        " bind_name": " None",
        " fast_count": " 4",
        " hold_multiplier": " 2",
        " interval": " 8",
        " management_address": " None",
        " mdn_notification_interval": " 5",
        " mdnstatus": " rxOnly",
        " notification_interval": " 5",
        " restart_delay": " 10",
        " transmit_delay": " 2",
        "lldpenable": " enabled"
    },
    "invocation": {
        "module_args": {
            "bind_name": null,
            "fast_count": 3,
            "hold_multiplier": 2,
            "interval": 8,
            "lldpenable": null,
            "management_address": null,
            "mdn_notification_interval": 10,
            "mdnstatus": null,
            "notification_interval": 10
            "restart_delay": 10,
            "state": "present",
            "transmit_delay": 1,
        }
    },
    "proposed": {
        "mdn_notification_interval": 10
    },
    "updates": [
        "lldp trap-interval 10"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
### show on device
```
[~8850_130.12]dis cu | in lldp
lldp enable
lldp mdn enable
lldp transmit interval 8
lldp transmit delay 1
lldp transmit multiplier 2
lldp restart 10
lldp trap-interval 10
lldp fast-count 3
lldp mdn trap-interval 10
[~8850_130.12]

```
### task 
```

  - name: "test the correct management_address"
    ce_lldp:
      bind_name: vlanif3111
```
### test log info
```
changed: [10.190.121.125] => {
    "changed": true,
    "end_state": {
        " bind_name": " Vlanif3111",
        " fast_count": " 3",
        " hold_multiplier": " 2",
        " interval": " 8",
        " management_address": " None",
        " mdn_notification_interval": " 10",
        " mdnstatus": " rxOnly",
        " notification_interval": " 10",
        " restart_delay": " 10",
        " transmit_delay": " 1",
        "lldpenable": " enabled"
    },
    "existing": {
        " bind_name": " None",
        " fast_count": " 3",
        " hold_multiplier": " 2",
        " interval": " 8",
        " management_address": " None",
        " mdn_notification_interval": " 10",
        " mdnstatus": " rxOnly",
        " notification_interval": " 10",
        " restart_delay": " 10",
        " transmit_delay": " 1",
        "lldpenable": " enabled"
    },
    "invocation": {
        "module_args": {
            "bind_name": "vlanif3111",
            "fast_count": null,
            "hold_multiplier": null,
            "interval": null,
            "lldpenable": null,
            "management_address": null,
            "mdn_notification_interval": null,
            "mdnstatus": null,
            "notification_interval": null,
            "restart_delay": null,
            "ssh_keyfile": null,
            "state": "present",
            "timeout": null,
            "transmit_delay": null
        }
    },
    "proposed": {
        "bind_name": "vlanif3111"
    },
    "updates": [
        " lldp management-address bind interface vlanif3111"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
### show configuration on device
```
[~8850_130.12]dis cu | in lldp
lldp enable
lldp mdn enable
lldp transmit interval 8
lldp transmit delay 1
lldp transmit multiplier 2
lldp restart 10
lldp trap-interval 10
lldp fast-count 3
lldp mdn trap-interval 10
lldp management-address bind interface Vlanif3111
[~8850_130.12]
```

### CODE REVIEW

**Comment by hwDCN** [comment]
LGTM

**Comment by hwDCN** [comment]
shipit

**Comment by zhongjun2** [comment]
LGTM

File: lib/ansible/modules/network/cloudengine/ce_lldp.py [hunk 1]
```diff
@@ -0,0 +1,788 @@
...
+                "transmit_delay": "4",
+                "notification_interval": "6",
+                "fast_count": "5",
+                "mdn_notification_interval": "6",
+                "management_address": "10.1.0.1",
+                "bind_name": "LoopBack2",
+                "state": "present"
+            }
+existing:
+    description: k/v pairs of existing global LLDP configration
```
Comment (Andersson007) [inline lib/ansible/modules/network/cloudengine/ce_lldp.py]:
```suggestion
    description: k/v pairs of existing global LLDP configuration
```

File: lib/ansible/modules/network/cloudengine/ce_lldp.py [hunk 2]
```diff
@@ -0,0 +1,788 @@
...
+    returned: always
+    type: dict
+    sample: {
+                "lldpenable": "disabled",
+                "mdnstatus": "disabled"
+            }
+end_state:
+    description: k/v pairs of global LLDP configration after module execution
```
Comment (Andersson007) [inline lib/ansible/modules/network/cloudengine/ce_lldp.py]:
```suggestion
    description: k/v pairs of global LLDP configuration after module execution
```

File: lib/ansible/modules/network/cloudengine/ce_lldp.py [hunk 3]
```diff
@@ -0,0 +1,788 @@
...
+CE_NC_MERGE_GLOBAL_LLDP_CONFIG_TAIL = """
+            </lldpSysParameter>
+        </lldpSys>
+    </lldp>
+</config>
+"""
+
+
+class Lldp(object):
+    """Manage global lldp enable configration"""
```
Comment (Andersson007) [inline lib/ansible/modules/network/cloudengine/ce_lldp.py]:
```suggestion
    """Manage global lldp enable configuration"""
```

File: lib/ansible/modules/network/cloudengine/ce_lldp.py [hunk 4]
```diff
@@ -0,0 +1,788 @@
...
+                self.module.fail_json(
+                    msg='Error: The value of mdn_notification_interval is out of [5 - 3600].')
+
+        if self.management_address:
+            self.is_valid_v4addr()
+
+        if self.bind_name:
+            if (len(self.bind_name) < 1) or (len(self.bind_name) > 63):
+                self.module.fail_json(
+                    msg='Error: Bind_name length is beetween 1 and 63.')
```
Comment (Andersson007) [inline lib/ansible/modules/network/cloudengine/ce_lldp.py]:
```suggestion
                    msg='Error: Bind_name length is between 1 and 63.')
```

File: lib/ansible/modules/network/cloudengine/ce_lldp.py [hunk 5]
```diff
@@ -0,0 +1,788 @@
...
+                        self.updates_cmd.append("lldp fast-count %s" % self.fast_count)
+                    if self.mdn_notification_interval:
+                        self.updates_cmd.append("lldp mdn trap-interval %s" % self.mdn_notification_interval)
+                    if self.management_address:
+                        self.updates_cmd.append("lldp management-address %s" % self.management_address)
+                    if self.bind_name:
+                        self.updates_cmd.append("lldp management-address bind interface %s" % self.bind_name)
+
+    def work(self):
+        """Excute task"""
```
Comment (Andersson007) [inline lib/ansible/modules/network/cloudengine/ce_lldp.py]:
```suggestion
        """Execute task"""
```

**Review by Andersson007 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_lldp.py (hunk 4 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_lldp.py]: Done.

File: lib/ansible/modules/network/cloudengine/ce_lldp.py (hunk 2 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_lldp.py]: Done.

File: lib/ansible/modules/network/cloudengine/ce_lldp.py (hunk 1 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_lldp.py]: Done.

**Comment by xuxiaowei0512** [comment]
@Andersson007  Done, it is so good for me.
Thank you so much.

**Comment by xuxiaowei0512** [comment]
read_for_review

**Review by yuandongxx — APPROVED** [review]
LGTM

**Comment by gundalow** [comment]
bot_status

**Comment by hwDCN** [comment]
shipit

**Comment by hwDCN** [comment]
bot_status

**Comment by hwDCN** [comment]
+1

**Comment by hwDCN** [comment]
LGTM

**Comment by xuxiaowei0512** [comment]
bot_status

**Comment by Andersson007** [comment]
!needs_revision

**Review by Andersson007 — APPROVED** [review]
The changes I requested were done correctly. I cannot comment on anything else.
(and i am far from the topic)

**Comment by xuxiaowei0512** [comment]
@Andersson007 @gundalow  @zhongjun2 
Thank you for your help.

**Pull request merged**